Add various hamlet features to xml-hamlet

dylex opened the pull request:
This adds support for fancy `bindings <-`, `$case`, and `*{attributes}` to xml-hamlet.  (The docs suggest these should exist already.)  I basically went through and merged pieces of Text.Hamlet.Parser into XMLParser, and Text.Hamlet into XML.  I believe I managed to avoid all the pieces that are specific to HTML/don't apply to XML.  Added a few test cases to match (though nothing near what hamlet has).  In my mind, the questionable pieces are:

- `ToAttributes`, for which I debated using the existing Text.Hamlet class, though that would disallow XML Names, and
- the overall duplication of code between packages (though this was a problem already).

I should also note that this is copying MIT-licensed code into BSD3, though @snoyberg owns both.

snoyberg commented:
Thanks for doing this. This is a large PR and will take me some time to review. That said: I'm no longer actively using or maintaining xml-hamlet. If you'd be interested in stepping in as a comaintainer, I'd be happy to share upload rights to Hackage if you'd like to maintain this going forward.

A cursory review of the patch looks good BTW.

dylex commented:
Thanks! I had never used xml-hamlet before yesterday, so I'm not sure I'm ready to take on maintainance, but if I do continue to find uses for it I will certainly consider it.  (My coworker new to haskell did say "[XML template] looks better than java/ruby analogs", so there's hope.)  Happy to iterate if you have any comments on this PR otherwise.

snoyberg commented:
I just tested this out on a piece of XML hamlet code I wrote recently:

```
                <p style="margin-bottom:2m" class="text-center">
                  <a class="btn btn-primary" role="button" data-toggle="collapse" href="#feedback-form#{ident}" aria-expanded="false" aria-controls="feedback-form">
                    #{buttonText}

                <div id="feedback-form#{ident}" class="collapse row">
                  <div class="col-sm-2">
                  <form method="post" action="/feedback" class="form-horizontal col-sm-8">

                    <div class="form-group">
                      <label for="feedback-form#{ident}-email">Email
                      <input id="feedback-form#{ident}-email" type="email" name="email" placeholder="Email" class="form-control">

                    <div class="form-group">
                      <label for="feedback-form#{ident}-name">Name
                      <input id="feedback-form#{ident}-name" type="text" name="name" placeholder="Name" class="form-control">

                    <div class="form-group">
                      <label for="feedback-form#{ident}-message">Message
                      <textarea id="feedback-form#{ident}-message" class="form-control" name="message" placeholder=#{message}>

                    <div class="form-group">
                      <input type="hidden" name="subject" value=#{message}>
                      <button type="submit" class="btn btn-success">#{submitText}

                    <p>
                      Or you can email us at #
                      <a href="mailto:#{dest}?subject=#{subject}">#{dest}
```

I received the error message:

```
(line 3, column 83):
unexpected "h"
expecting ">"
```

This looks related to https://github.com/yesodweb/shakespeare/issues/200, which unfortunately was a bug that slipped in just before you did this refactoring I think.

dylex commented:
Updated to match shakespeare/2.0.12.1 (two line change at XMLParse.hs:190), added matching test case, and added specific version in commit message, to help with future merging.  I'll now watch the shakespeare repo for future updates as well.

snoyberg commented:
Thanks for the quick turnaround here! Just doing a few more sanity checks.

snoyberg commented:
Thank you!
